### PR TITLE
Updated: Resolved all Issues.

### PR DESCRIPTION
**Pull request by Goutam-n**
- Updated README.md
- Updated services

### CODE REVIEW

File: service/mantle/ShipEngineServices.xml [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    <service verb="get" noun="ShippingRate">
+        <!-- see: https://www.shipengine.com/docs/rates/#example-with-shipment-details -->
+        <implements service="mantle.shipment.CarrierServices.get#ShippingRate"/>
+        <in-parameters>
+            <parameter name="shippingGatewayConfigId" default-value="SHIPENGINE_DEMO"/>
+            <parameter name="shipmentRouteSegmentSeqId" default-value="01"/>
+            <parameter name="shipmentPackageSeqId" default-value="01"/>
+        </in-parameters>
+        <out-parameters>
+            <parameter name="requestMap"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need to out request map.

File: service/mantle/ShipEngineServices.xml [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            <!-- ==============  Shipment_Record  ============== -->
+
+            <entity-find-one entity-name="mantle.shipment.Shipment" value-field="shipment" for-update="true"/>
+            <if condition="shipment == null"><return error="true" message="Shipment not found with ID ${shipmentId}"/></if>
+            <entity-find-one entity-name="mantle.shipment.ShipmentRouteSegment" value-field="routeSegment"/>
+            <if condition="routeSegment == null"><return error="true" message="Shipment Route Segment not found with ID ${shipmentId}:${shipmentRouteSegmentSeqId}"/></if>
+            <set field="originPostalContactMechId" from="routeSegment.originPostalContactMechId"/>
+            <set field="destPostalContactMechId" from="routeSegment.destPostalContactMechId"/>
+            <set field="hasIssues" from="false"/>
+            <set field="fromPartyId" from="shipment.fromPartyId"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: We are not using fromPartyId and toPartyId anywhere in the code. Remove these unnecessary variables.

File: service/mantle/ShipEngineServices.xml [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+            <if condition="!toPartyId"><set field="hasIssues" from="true"/><message type="danger">To party not set on shipment ${shipmentId}</message></if>
+            <if condition="hasIssues"><message type="danger">Issues found, not requesting Shippo labels</message><return/></if>
+<!--            <log level="info" message="==================${routeSegment}========================="/>-->
+            <if condition="!originPostalContactMechId &amp;&amp; routeSegment.originFacilityId">
+                <service-call name="mantle.facility.ContactServices.get#FacilityDefaultShipOrigin" out-map="facOriginOut"
+                              in-map="[facilityId:routeSegment.originFacilityId]"/>
+                <if condition="facOriginOut.postalContactMechId">
+                    <set field="originPostalContactMechId" from="facOriginOut.postalContactMechId"/>
+                    <set field="routeSegment.originPostalContactMechId" from="originPostalContactMechId"/>
+                    <entity-update value-field="routeSegment"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need to update route segment.

File: service/mantle/ShipEngineServices.xml [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+            <entity-find-one entity-name="mantle.shipment.ShipmentRouteSegment" value-field="routeSegment"/>
+            <if condition="routeSegment == null"><return error="true" message="Shipment Route Segment not found with ID ${shipmentId}:${shipmentRouteSegmentSeqId}"/></if>
+            <set field="originPostalContactMechId" from="routeSegment.originPostalContactMechId"/>
+            <set field="destPostalContactMechId" from="routeSegment.destPostalContactMechId"/>
+            <set field="hasIssues" from="false"/>
+            <set field="fromPartyId" from="shipment.fromPartyId"/>
+            <if condition="!fromPartyId"><set field="hasIssues" from="true"/><message type="danger">From party not set on shipment ${shipmentId}</message></if>
+            <set field="toPartyId" from="shipment.toPartyId"/>
+            <if condition="!toPartyId"><set field="hasIssues" from="true"/><message type="danger">To party not set on shipment ${shipmentId}</message></if>
+            <if condition="hasIssues"><message type="danger">Issues found, not requesting Shippo labels</message><return/></if>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: Remove this validation.

File: service/mantle/ShipEngineServices.xml [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            <if condition="!originPostalContactMechId &amp;&amp; routeSegment.originFacilityId">
+                <service-call name="mantle.facility.ContactServices.get#FacilityDefaultShipOrigin" out-map="facOriginOut"
+                              in-map="[facilityId:routeSegment.originFacilityId]"/>
+                <if condition="facOriginOut.postalContactMechId">
+                    <set field="originPostalContactMechId" from="facOriginOut.postalContactMechId"/>
+                    <set field="routeSegment.originPostalContactMechId" from="originPostalContactMechId"/>
+                    <entity-update value-field="routeSegment"/>
+                </if>
+            </if>
+            <if condition="!routeSegment.carrierPartyId"><log level="warn" message="This ShipmentID ===${shipmentId}=== has no carrierPartyId, not getting rates"/></if>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need for this validation.

File: service/mantle/ShipEngineServices.xml [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+            <if condition="!serviceCode"><log level="warn" message="Shipping gateway ===${shippingGatewayConfigId}=== has no gatewayServiceCode, not getting rates"/><return/></if>
+
+
+            <!-- ==============  Carrier_Id  ============== -->
+
+            <entity-find-one entity-name="mantle.shipment.carrier.ShippingGatewayCarrier" value-field="gatewayCarrier">
+                <field-map field-name="shippingGatewayConfigId"/>
+                <field-map field-name="carrierPartyId" from="routeSegment.carrierPartyId"/>
+            </entity-find-one>
+            <set field="carrierIds" type="List" from="[gatewayCarrier.gatewayAccountId]"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml:63]: This shouldn't be a list.

File: service/mantle/ShipEngineServices.xml [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+            <if condition="!carrierIds"><log level="warn" message="Shipping Gateway ${shippingGatewayConfigId} has no carrierId, not getting rates"/><return/></if>
+
+
+            <!-- ==============  Address_to  ============== -->
+
+            <if condition="!destPostalContactMechId &amp;&amp; routeSegment.destinationFacilityId">
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: We don't need line 81-89.

File: service/mantle/ShipEngineServices.xml [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+                </entity-find-one>
+                <entity-find-one entity-name="mantle.party.contact.TelecomNumber" value-field="destTelecomNumber">
+                    <field-map field-name="contactMechId" from="destAddress.telecomContactMechId"/>
+                </entity-find-one>
+                <set field="number" from="destTelecomNumber.areaCode+'-'+destTelecomNumber.contactNumber"/>
+                <set field="shipTo" from="[name:destAddress.toName,phone:number,address_line1:destAddress.address1,
+                       city_locality: destAddress.city,state_province:destState.geoCodeAlpha2,postal_code:destAddress.postalCode,
+                       country_code:destCountry.geoCodeAlpha2,address_residential_indicator:indicatorTo ]"/>
+            </then><else>
+                <message error="true">No destination address set on Shipment Route Segment ${shipmentId}:${shipmentRouteSegmentSeqId}</message>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need of this validation.

File: service/mantle/ShipEngineServices.xml [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+            </else></if>
+
+
+            <!-- ==============  Address_From  ============== -->
+
+            <if condition="!originPostalContactMechId &amp;&amp; routeSegment.originFacilityId">
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml:100]: This code is repetitive.

File: service/mantle/ShipEngineServices.xml [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+                    <field-map field-name="geoId" from="originAddress.countryGeoId"/>
+                </entity-find-one>
+                <entity-find-one entity-name="mantle.party.contact.TelecomNumber" value-field="originTelecomNumber">
+                    <field-map field-name="contactMechId" from="originAddress.telecomContactMechId"/>
+                </entity-find-one>
+                <set field="number" from="originTelecomNumber.areaCode+'-'+originTelecomNumber.contactNumber"/>
+                <set field="shipFrom" from="[name:originAddress.toName,phone:number,address_line1:originAddress.address1,
+                       city_locality: originAddress.city, state_province:originState.geoCodeAlpha2,postal_code:originAddress.postalCode,
+                       country_code:originCountry.geoCodeAlpha2,address_residential_indicator:indicatorFrom]"/>
+            </then><else>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need of this validation.

File: service/mantle/ShipEngineServices.xml [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+            <set field="responseMap" from="[]"/>
+            <set field="errMsg" type="String"/>
+
+            <script><![CDATA[
+                    org.moqui.util.RestClient restClient = ec.service.rest().method(org.moqui.util.RestClient.POST)
+                    .addHeader("API-Key", "${apiToken}")
+                    .addHeader("Content-Type", "application/json").jsonObject(requestMap)
+                    restClient.uri().protocol("https").host("api.shipengine.com").port(443).path("v1/rates").build()
+                    org.moqui.util.RestClient.RestResponse restResponse = restClient.call()
+                    if (restResponse.statusCode < 200 || restResponse.statusCode >= 300) {
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml:176]: Replace with,
`
if (restResponse.statusCode < 200 || restResponse.statusCode >= 300) {
    ec.logger.warn("Unsuccessful with status code: ${statusCode} and response: ${response}")
    return
}
`

File: service/mantle/ShipEngineServices.xml [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+        <implements service="mantle.shipment.CarrierServices.get#ShippingRate"/>
+        <in-parameters>
+            <parameter name="shippingGatewayConfigId" default-value="SHIPENGINE_DEMO"/>
+            <parameter name="shipmentRouteSegmentSeqId" default-value="01"/>
+            <parameter name="shipmentPackageSeqId" default-value="01"/>
+        </in-parameters>
+        <out-parameters>
+            <parameter name="requestMap"/>
+            <parameter name="responseMap"/>
+            <parameter name="errMsg"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: No need of this variable.

File: service/mantle/ShipEngineServices.xml (hunk 1 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issue successfully.

File: service/mantle/ShipEngineServices.xml (hunk 3 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 4 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 7 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 9 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:85]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 10 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 11 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: service/mantle/ShipEngineServices.xml (hunk 12 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Resolved Issues successfully.

File: component.xml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://moqui.org/xsd/moqui-conf-3.xsd"
+           name="mantle-shipengine" version="1.2.0">
```
Comment (mridulpathak) [inline component.xml:3]: Formatting issue here.

File: data/ShipEngineDemoData.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    <mantle.shipment.ShipmentBoxType shipmentBoxTypeId="USPS_MdFlat1" description="USPS Medium Flat Rate Box 1" dimensionUomId="LEN_in" boxLength="11.25" boxWidth="8.75" boxHeight="6" gatewayBoxId="USPS_MediumFlatRateBox1"/>
+    <mantle.shipment.ShipmentBoxType shipmentBoxTypeId="USPS_LgFlat" description="USPS Large Flat Rate Box" dimensionUomId="LEN_in" boxLength="12.25" boxWidth="12.25" boxHeight="6" gatewayBoxId="USPS_LargeFlatRateBox"/>
+
+    <mantle.shipment.carrier.ShippingGatewayConfig shippingGatewayConfigId="SHIPENGINE_DEMO"
+                                                   shippingGatewayTypeEnumId="ShGtwyShipEngine" description="ShipEngine API Demo"
+                                                   getRateServiceName="mantle.ShipEngine.ShipEngineServices.get#ShippingRate"
+                                                   requestLabelsServiceName="mantle.ShipEngineServices.post#CreateLabel">
+
+        <!--  set your ShipEngine Account IDs for carriers with records like these  -->
+        <carriers carrierPartyId="UPS" gatewayAccountId="se-2084091"/>
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml]: Keep gatewayAccountId as empty.

File: data/ShipEngineDemoData.xml [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        <carriers carrierPartyId="USPS" gatewayAccountId="se-2084090"/>
+
+        <!--  set your ShipEngine API token (test or live) with a record like this  -->
+        <options optionEnumId="SgoApiToken" optionValue=""/>
+        <options optionEnumId="SgoLabelType" optionValue="PDF_4x6"/>
+
+        <!--  use SCopyhippingGatewayBoxType to override default gatewayBoxId on ShipmentBoxType  -->
+        <boxTypes shipmentBoxTypeId="USPS_SmFlat" gatewayBoxId="USPS_SmallFlatRateBox"/>
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml]: Do we need this?

File: data/ShipEngineDemoData.xml [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+        <!--  use ShippingGatewayMethod to override default gatewayServiceCode on CarrierShipmentMethod  -->
+        <methods carrierPartyId="USPS" shipmentMethodEnumId="ShMthGround" gatewayServiceCode="usps_first_class_mail"/>\
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml]: Make sure that gatewayServiceCode mapping here is correct for USPS Ground.

File: data/ShipEngineDemoData.xml [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        <methods carrierPartyId="UPS" shipmentMethodEnumId="ShMthGround" gatewayServiceCode="ups_ground"/>
+    </mantle.shipment.carrier.ShippingGatewayConfig>
+
+    <!--  used in mantle.shipment.CarrierServices ShipmentLabels services  -->
+    <moqui.security.UserGroupPreference userGroupId="ALL_USERS" preferenceKey="ValidateAddressGatewayConfigId" preferenceValue="SHIPENGINE_DEMO"/>
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml]: We haven't implemented address validation service, do we need this?

File: data/ShipEngineDemoData.xml [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    <moqui.security.UserGroupPreference userGroupId="ALL_USERS" preferenceKey="DefaultShipmentGatewayConfigId" preferenceValue="SHIPENGINE_DEMO"/>
+
+    <!--  example per org/vendor settings, used as defaults if not store; note that for screens like EditParty with no store or
+            vendor the Owner Party must be set to use these settings  -->
+    <mantle.party.PartySetting partyId="ORG_ZIZI_RETAIL" partySettingTypeId="ValidateAddressGatewayConfigId" settingValue="SHIPENGINE_DEMO"/>
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml]: Same as above.

File: data/ShipEngineDemoData.xml [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    <mantle.party.PartySetting partyId="ORG_ZIZI_RETAIL" partySettingTypeId="DefaultShipmentGatewayConfigId" settingValue="SHIPENGINE_DEMO"/>
+
+    <mantle.product.store.ProductStore productStoreId="POPC_DEFAULT">
```
Comment (mridulpathak) [inline data/ShipEngineDemoData.xml:32]: Doesn't match with ShipmentGatewayMethod above.

File: service/mantle/ShipEngineServices.xml [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+
+            <!-- ==============  API-Key Authentication  ============== -->
+
+            <entity-find-one entity-name="mantle.shipment.carrier.ShippingGatewayOption" value-field="apiTokenOpt">
+                <field-map field-name="shippingGatewayConfigId"/>
+                <field-map field-name="optionEnumId" value="SgoApiToken"/>
+            </entity-find-one>
+            <set field="apiToken" from="apiTokenOpt?.optionValue"/>
+            <if condition="!apiToken">
+                <log level="warn" message="Shipping gateway config ${shippingGatewayConfigId} has no SgoApiToken, not getting rates"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: log level should be error.

File: service/mantle/ShipEngineServices.xml [hunk 21]
```diff
@@ -0,0 +1,453 @@
...
+                <return/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: return with an error message.

File: service/mantle/ShipEngineServices.xml [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+                ]]>
+            </script>
+        </actions>
+    </service>
+
+    <!-- ============================================= -->
+    <!-- ========== Create Label Service ============= -->
+    <!-- ============================================= -->
+
+    <service verb="post" noun="CreateLabel">
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: verb="create" noun="ShippingLabel"

File: data/ShipEngineDemoData.xml (hunk 14 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:13]: Resolved Issues successfully.

File: data/ShipEngineDemoData.xml (hunk 15 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:21]: Resolved Issues successfully.

File: data/ShipEngineDemoData.xml (hunk 16 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:24]: Changed gatewayServiceCode

File: data/ShipEngineDemoData.xml (hunk 17 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:29]: Issue resolved successfully.

File: data/ShipEngineDemoData.xml (hunk 18 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:34]: Issue resolved successfully.

File: data/ShipEngineDemoData.xml (hunk 19 above)
Comment (Goutam-n) [reply data/ShipEngineDemoData.xml:37]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml (hunk 20 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:24]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml (hunk 21 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:25]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml (hunk 22 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:191]: Changed verb

File: service/mantle/ShipEngineServices.xml [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+            </entity-find-one>
+            <if condition="!gatewayMethod">
+                <log level="error" message="Shipping gateway config ${shippingGatewayConfigId} has no ShippingGatewayMethod, not getting rates"/>
+                <return/>
+            </if>
+            <if condition="!serviceCode">
+                <log level="error" message="Shipping gateway config ${shippingGatewayConfigId} has no gatewayServiceCode, not getting rates"/>
+                <return/>
+            </if>
+            <set field="serviceCode" from="[gatewayMethod.gatewayServiceCode]"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: error check is placed wrongly.

File: service/mantle/ShipEngineServices.xml [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+            </if>
+            <entity-find-one entity-name="moqui.basic.UomConversion" value-field="conversion">
+                <field-map field-name="uomId" from="shipmentPackage.weightUomId"/>
+                <field-map field-name="toUomId" value="WT_oz"/>
+            </entity-find-one>
+            <if condition="!conversion">
+                <log level="error" message="This ShipmentId ${shipmentId} has no weight in UomConversion, not getting rates"/>
+                <return/>
+            </if>
+            <set field="convertedWeight" from="shipmentPackage.weight*conversion.conversionFactor"/>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml:160]: There should be validation for shipment shipmentPackage.weight.

File: service/mantle/ShipEngineServices.xml (hunk 23 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:54]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml (hunk 24 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml:171]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml [hunk 25]
```diff
@@ -0,0 +1,476 @@
...
+                ]]>
+            </script>
+        </actions>
+    </service>
+
+    <!-- ============================================= -->
+    <!-- ========== Create Label Service ============= -->
+    <!-- ============================================= -->
+
+    <service verb="create" noun="CreateLabel">
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: CreateLabel >> ShippingLabel

File: service/mantle/ShipEngineServices.xml [hunk 26]
```diff
@@ -0,0 +1,476 @@
...
+
+            <entity-update value-field="packageRouteSeg"/>
+        </actions>
+    </service>
+
+    <!-- ============================================= -->
+    <!-- ========== Void Label Service =============== -->
+    <!-- ============================================= -->
+
+    <service verb="put" noun="VoidLabel">
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: verb="void"
noun="ShippingLabel"

File: README.md [hunk 27]
```diff
@@ -1,2 +1,14 @@
 # mantle-shipengine
 ShipEngine
+Integration with ShipEngine for shipping rates, create labels, and void label. In this Integration we worked on staging environment. So, the rates may vary from production evironment.
```
Comment (mridulpathak) [inline README.md:3]: Readme file is still not updated as per suggestions.

File: service/mantle/ShipEngineServices.xml [hunk 28]
```diff
@@ -0,0 +1,452 @@
...
+
+            <!-- ==============  Address_to  ============== -->
+
+            <entity-find-one entity-name="mantle.party.contact.PostalAddress" value-field="destAddress">
+                <field-map field-name="contactMechId" from="routeSegment.destPostalContactMechId"/>
+            </entity-find-one>
+            <if condition="!destAddress">
+                <return error="true" message="This contactMechId ${contactMechId} has no destination address, not getting rates"/>
+            </if>
+            <script>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: This can be done in <set> tag, no need of script.

File: service/mantle/ShipEngineServices.xml [hunk 29]
```diff
@@ -0,0 +1,452 @@
...
+                </if>
+            </if>
+
+            <entity-find-one entity-name="mantle.party.contact.PostalAddress" value-field="originAddress">
+                <field-map field-name="contactMechId" from="routeSegment.originPostalContactMechId"/>
+            </entity-find-one>
+            <if condition="!originAddress">
+                <return error="true" message="This contactMechId ${contactMechId} has no origin address, not getting rates"/>
+            </if>
+            <script>
```
Comment (mridulpathak) [inline service/mantle/ShipEngineServices.xml]: Same as above.

File: component.xml [hunk 30]
```diff
@@ -0,0 +1,5 @@
...
+    <depends-on name="mantle-usl" version="2.2.0"/>
```
Comment (mridulpathak) [inline component.xml:4]: I don't see any OOTB service being used in this code so we shouldn't be needing this dependency. In fact we are using OOTB data model, so dependency should be added for mantle-udm instead. Or may be just leave a comment.

File: service/mantle/ShipEngineServices.xml (hunk 25 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Issue resolved successfully.

File: service/mantle/ShipEngineServices.xml (hunk 26 above)
Comment (Goutam-n) [reply service/mantle/ShipEngineServices.xml]: Issue resolved successfully.

File: README.md (hunk 27 above)
Comment (Goutam-n) [reply README.md:3]: Issue resolved successfully.

File: README.md [hunk 31]
```diff
@@ -1,2 +1,17 @@
...
+
+To use simply:
+
+Load the setup data in data/ShipEngineSetupData.xml
+Load the demo configuration data in data/ShipEngineDemoData.xml or create your own configuration and load it; if you use the demo data, add your API token (SgoApiToken).
+
+~ get#ShippingRates service: 
+This API returns a rate as per the destination location and parcel attributes.
+
+~ create#CreateLabel service :
```
Comment (mridulpathak) [inline README.md]: service name needs to be updated.

File: README.md [hunk 32]
```diff
@@ -1,2 +1,17 @@
...
+This label API will return a URL to a label for an order.
+
+~ put#VoidLabel service:
```
Comment (mridulpathak) [inline README.md]: service name needs to be updated.

File: README.md (hunk 31 above)
Comment (Goutam-n) [reply README.md]: Issue resolved successfully.

**Pull request merged**